### PR TITLE
[FEATURE] Demote abandoned packages in template list

### PR DESCRIPTION
**Pull request by mteu**
In the `PackagistProvider` we're currently showing packages based on the package type ignoring the state the published package is in. The could result in showing packages that are marked as abandoned by the maintainers. This PR seeks to disclaim this issue by sorting abadoned packages to the bottom of the list and adding a disclaimer to them.

I propose a rather minimal invasive disclaimer for the time being.

grafik

Related: #243
Co-authored-by: @eliashaeussler

### CODE REVIEW

File: tests/src/Template/Provider/BaseProviderTest.php [hunk 1]
```diff
@@ -276,15 +292,25 @@ private static function createPackage(
         string $name,
         string $type = 'project-builder-template',
         string $prettyVersion = '1.0.0',
-    ): Package\Package {
+    ): Package\CompletePackage {
         $versionParser = new VersionParser();
 
-        $package = new Package\Package($name, $versionParser->normalize($prettyVersion), $prettyVersion);
+        $package = new Package\CompletePackage($name, $versionParser->normalize($prettyVersion), $prettyVersion);
```
Comment (mteu) [inline tests/src/Template/Provider/BaseProviderTest.php:298]: @eliashaeussler Please let me know if that's too invasive. I'd revert this change and refactor `createAbandonedPackage()` accordingly.

**Comment by eliashaeussler** [comment]
@mteu Thanks a bunch for this PR. I tried it locally and think that it's not intrusive enough to attract users' attention. What about your original implementation? I think I like that approach better, because it's clearer to the user what's actually going on and cannot be overlooked.

image

**Comment by eliashaeussler** [comment]
In addition, we may think about a config option to hide packages from the list of template sources. Something like this in `composer.json`:

```json
{
    "extra": {
        "cpsit/project-builder": {
            "hide-from-template-sources": true
        }
    }
}
```

File: src/Template/Provider/BaseProvider.php [hunk 2]
```diff
@@ -87,15 +87,40 @@ public function listTemplateSources(): array
             self::PACKAGE_TYPE,
         );
 
-        foreach ($searchResult as ['name' => $packageName]) {
-            $package = $repository->findPackage($packageName, $constraint);
+        foreach ($searchResult as $result) {
+            $package = $repository->findPackage($result['name'], $constraint);
```
Comment (eliashaeussler) [inline src/Template/Provider/BaseProvider.php:91]: Can be reverted.

File: src/Template/Provider/BaseProvider.php [hunk 3]
```diff
@@ -14,7 +14,7 @@
  *
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
```
Comment (eliashaeussler) [inline src/Template/Provider/BaseProvider.php:17]: Can be reverted.

File: tests/src/Template/Provider/BaseProviderTest.php [hunk 4]
```diff
@@ -276,15 +292,25 @@ private static function createPackage(
...
+    private static function createAbandonedPackage(
+        string $name,
+        bool|string $abandoned,
+    ): Package\CompletePackage {
+        $package = self::createPackage($name);
+        $package->setAbandoned($abandoned);
+
+        return $package;
+    }
+
```
Comment (eliashaeussler) [inline tests/src/Template/Provider/BaseProviderTest.php:313]: Let's not create an additional method for abandoned packages, but rather use the existing `createPackage()` method with an additional property `bool $abandoned = false`.

File: src/Template/Provider/BaseProvider.php [hunk 5]
```diff
@@ -87,15 +87,40 @@ public function listTemplateSources(): array
...
+                && $templateSource->getPackage()->isAbandoned()
+            ) {
+                $abandonedPackages[] = $templateSource;
+                continue;
+            }
+
+            $orderedTemplateSources[] = $templateSource;
+        }
+
+        return array_merge($orderedTemplateSources, $abandonedPackages);
```
Comment (eliashaeussler) [inline src/Template/Provider/BaseProvider.php:123]: What about doing the whole merging right after the `foreach` in `listTemplateSources()` instead of creating a new method for this? This would avoid iterating twice over the found packages. Within `listTemplateSources()`, we could create two variables `$maintainedPackages` and `$abandonedPackages` and do the array merge right after the `foreach`, just as you did here in `sortAbandonedPackagesToBottomOfTemplateSources()`.

File: src/IO/Messenger.php [hunk 6]
```diff
@@ -468,7 +468,28 @@ private function decorateTemplateSource(Template\TemplateSource $templateSource)
             return $name;
         }
 
-        return sprintf('%s <fg=gray>(%s)</>', $description, $name);
+        return sprintf(
```
Comment (eliashaeussler) [inline src/IO/Messenger.php:471]: This does not apply to packages without a description (see the early return on line 467).

File: src/IO/Messenger.php [hunk 7]
```diff
@@ -468,7 +468,28 @@ private function decorateTemplateSource(Template\TemplateSource $templateSource)
...
+
+        return sprintf(
+            '%s %s%s',
+            $package->getPrettyName(),
+            'is abandoned.',
+            match ($package->getReplacementPackage()) {
+                null => ' No replacement was suggested.',
+                default => ' Use '.$package->getReplacementPackage().' instead.',
+            },
+        );
```
Comment (eliashaeussler) [inline src/IO/Messenger.php:492]: See my comment in the PR. I'd suggest the following, if we choose the other output:

```suggestion
        $decoratedTemplateSource = sprintf('%s <fg=gray>(%s)</>', $description, $name);

        if (!$package->isAbandoned()) {
            return $decoratedTemplateSource;
        }

        return sprintf(
            '%s <warning>Abandoned! %s</warning>',
            $decoratedTemplateSource,
            match ($package->getReplacementPackage()) {
                null => 'No replacement was suggested.',
                default => 'Use '.$package->getReplacementPackage().' instead.',
            },
        );
```

File: tests/src/Template/Provider/BaseProviderTest.php (hunk 1 above)
Comment (eliashaeussler) [reply tests/src/Template/Provider/BaseProviderTest.php:298]: Fine for me.

**Review by eliashaeussler — CHANGES_REQUESTED** [review]

File: src/Template/Provider/BaseProvider.php (hunk 2 above)
Comment (mteu) [reply src/Template/Provider/BaseProvider.php:91]: Perfectly right. 🙇

File: tests/src/Template/Provider/BaseProviderTest.php (hunk 4 above)
Comment (mteu) [reply tests/src/Template/Provider/BaseProviderTest.php]: Again, completely agree. I've refactored accordingly.

File: src/Template/Provider/BaseProvider.php (hunk 3 above)
Comment (mteu) [reply src/Template/Provider/BaseProvider.php:17]: Hm. I realise that you have these excessive spaces all over. But this topic seems out of scope. Reverting as requested.

Comment (eliashaeussler) [reply src/Template/Provider/BaseProvider.php]: Yeah, that's probably a bug in my live template :D Let's address this in another change.

File: src/IO/Messenger.php (hunk 6 above)
Comment (mteu) [reply src/IO/Messenger.php]: Nah, that's obviously a bug. I addressed in 01992b2c6b09e3ca9093c7ed384e8504f5ff857a.

**Review by eliashaeussler — APPROVED** [review]
LGTM 👍 Nice work here! Thanks a bunch. 🚀
